test/lib/commands/ci.js

```diff
@@ -230,3 +226,85 @@ t.test('should throw error when ideal inventory mismatches virtual', async t =>
   const nmTestFile = path.join(npm.prefix, 'node_modules', 'test-file')
   t.equal(fs.existsSync(nmTestFile), true, 'does not remove node_modules')
 })
+
+t.test('should remove dirty node_modules with unhoisted workspace module', async t => {
+  const { npm, registry, assert } = await loadMockNpm(t, {
+    prefixDir: workspaceMock(t, {
+      clean: false,
+      workspaces: {
+        'workspace-a': {
+          'abbrev@1.1.0': { },
+        },
+        'workspace-b': {
+          'abbrev@1.1.1': { hoist: false },
+        },
+      },
+    }),
+  })
+  await registry.setup({
+    'abbrev@1.1.0': path.join(npm.prefix, 'tarballs/abbrev@1.1.0'),
+    'abbrev@1.1.1': path.join(npm.prefix, 'tarballs/abbrev@1.1.1'),
+  })
+  registry.nock.post('/-/npm/v1/security/advisories/bulk').reply(200, {})
+  assert.packageDirty('node_modules/abbrev@1.1.0')
+  assert.packageDirty('workspace-b/node_modules/abbrev@1.1.1')
+  await npm.exec('ci', [])
+  assert.packageInstalled('node_modules/abbrev@1.1.0')
+  assert.packageInstalled('workspace-b/node_modules/abbrev@1.1.1')
+})
+
+t.test('should remove dirty node_modules with hoisted workspace modules', async t => {
+  const { npm, registry, assert } = await loadMockNpm(t, {
+    prefixDir: workspaceMock(t, {
+      clean: false,
+      workspaces: {
+        'workspace-a': {
+          'abbrev@1.1.0': { },
+        },
+        'workspace-b': {
+          'lodash@1.1.1': { },
+        },
+      },
+    }),
+  })
+  await registry.setup({
+    'abbrev@1.1.0': path.join(npm.prefix, 'tarballs/abbrev@1.1.0'),
+    'lodash@1.1.1': path.join(npm.prefix, 'tarballs/lodash@1.1.1'),
+  })
+  registry.nock.post('/-/npm/v1/security/advisories/bulk').reply(200, {})
+  assert.packageDirty('node_modules/abbrev@1.1.0')
+  assert.packageDirty('node_modules/lodash@1.1.1')
+  await npm.exec('ci', [])
+  assert.packageInstalled('node_modules/abbrev@1.1.0')
+  assert.packageInstalled('node_modules/lodash@1.1.1')
+})
+
+/** this behaves the same way as install but will remove all workspace node_modules */
+t.test('should use --workspace flag', async t => {
+  t.saveFixture = true
+  const { npm, registry, assert } = await loadMockNpm(t, {
+    config: {
+      workspace: 'workspace-b',
+    },
+    prefixDir: workspaceMock(t, {
+      clean: false,
+      workspaces: {
+        'workspace-a': {
+          'abbrev@1.1.0': { },
+        },
+        'workspace-b': {
+          'lodash@1.1.1': { },
+        },
+      },
+    }),
+  })
+  await registry.setup({
+    'lodash@1.1.1': path.join(npm.prefix, 'tarballs/lodash@1.1.1'),
+  })
+  registry.nock.post('/-/npm/v1/security/advisories/bulk').reply(200, {})
+  assert.packageDirty('node_modules/abbrev@1.1.0')
+  assert.packageDirty('node_modules/lodash@1.1.1')
+  await npm.exec('ci', [])
+  assert.packageMissing('node_modules/abbrev@1.1.0')
+  assert.packageInstalled('node_modules/lodash@1.1.1')
```

Comment:
I recently added convince assertion methods around checking before and after I think this makes it a lot easier easier to see what the commands does.

Reply:
I think they help a lot too, but I also think under the hood they should use tap's assertions instead of `node:assert`.  Let's stick to a single assertion library here.